File: lib/supabase.ts
import { createClient, SupabaseClient } from '@supabase/supabase-js'
import type { Report, AnalysisResult } from '@/types'

// Lazy-initialized clients (avoid crashes during build when env vars are empty)
let _supabaseClient: SupabaseClient | null = null
let _supabaseAdmin: SupabaseClient | null = null

export function getSupabaseClient(): SupabaseClient {
  if (!_supabaseClient) {
    _supabaseClient = createClient(
      process.env.NEXT_PUBLIC_SUPABASE_URL!,
      process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY!
    )
  }
  return _supabaseClient
}

function getSupabaseAdmin(): SupabaseClient {
  if (!_supabaseAdmin) {
    _supabaseAdmin = createClient(
      process.env.NEXT_PUBLIC_SUPABASE_URL!,
      process.env.SUPABASE_SERVICE_ROLE_KEY!
    )
  }
  return _supabaseAdmin
}

// ---------------------------------------------------------------------------
// Helpers: convert Supabase snake_case rows to camelCase Report
// ---------------------------------------------------------------------------

function toReport(row: Record<string, unknown>): Report {
  return {
    id: row.id as string,
    userEmail: row.user_email as string,
    productName: row.product_name as string | undefined,
    productDescription: row.product_description as string,
    imageUrl: row.image_url as string | undefined,
    targetPrice: row.target_price as string | undefined,
    mainConcern: row.main_concern as string | undefined,
    quantity: row.quantity as string | undefined,
    analysisResult: row.analysis_result as AnalysisResult | undefined,
    status: row.status as Report['status'],
    stripeSessionId: row.stripe_session_id as string | undefined,
    stripePaymentId: row.stripe_payment_id as string | undefined,
    isPaid: row.is_paid as boolean,
    plan: row.plan as Report['plan'],
    shareCount: row.share_count as number,
    createdAt: row.created_at as string,
    completedAt: row.completed_at as string | undefined,
  }
}

function toRow(data: Partial<Report>): Record<string, unknown> {
  const map: Record<string, unknown> = {}
  if (data.userEmail !== undefined) map.user_email = data.userEmail
  if (data.productName !== undefined) map.product_name = data.productName
  if (data.productDescription !== undefined) map.product_description = data.productDescription
  if (data.imageUrl !== undefined) map.image_url = data.imageUrl
  if (data.targetPrice !== undefined) map.target_price = data.targetPrice
  if (data.mainConcern !== undefined) map.main_concern = data.mainConcern
  if (data.quantity !== undefined) map.quantity = data.quantity
  if (data.analysisResult !== undefined) map.analysis_result = data.analysisResult
  if (data.status !== undefined) map.status = data.status
  if (data.stripeSessionId !== undefined) map.stripe_session_id = data.stripeSessionId
  if (data.stripePaymentId !== undefined) map.stripe_payment_id = data.stripePaymentId
  if (data.isPaid !== undefined) map.is_paid = data.isPaid
  if (data.plan !== undefined) map.plan = data.plan
  if (data.shareCount !== undefined) map.share_count = data.shareCount
  if (data.completedAt !== undefined) map.completed_at = data.completedAt
  return map
}

// ---------------------------------------------------------------------------
// CRUD Operations
// ---------------------------------------------------------------------------

export async function createReport(
  data: Omit<Report, 'id' | 'createdAt'>
): Promise<Report> {
  const { data: row, error } = await getSupabaseAdmin()
    .from('reports')
    .insert(toRow(data))
    .select()
    .single()

  if (error) {
    console.error('createReport error:', error)
    throw new Error(`Failed to create report: ${error.message}`)
  }
  return toReport(row)
}

export async function getReport(id: string): Promise<Report | null> {
  const { data: row, error } = await getSupabaseAdmin()
    .from('reports')
    .select()
    .eq('id', id)
    .single()

  if (error) {
    if (error.code === 'PGRST116') return null // not found
    console.error('getReport error:', error)
    throw new Error(`Failed to get report: ${error.message}`)
  }
  return toReport(row)
}

export async function updateReport(
  id: string,
  data: Partial<Report>
): Promise<Report> {
  const { data: row, error } = await getSupabaseAdmin()
    .from('reports')
    .update(toRow(data))
    .eq('id', id)
    .select()
    .single()

  if (error) {
    console.error('updateReport error:', error)
    throw new Error(`Failed to update report: ${error.message}`)
  }
  return toReport(row)
}

export async function markReportPaid(
  id: string,
  stripePaymentId: string
): Promise<void> {
  const { error } = await getSupabaseAdmin()
    .from('reports')
    .update({ is_paid: true, stripe_payment_id: stripePaymentId })
    .eq('id', id)

  if (error) {
    console.error('markReportPaid error:', error)
    throw new Error(`Failed to mark report as paid: ${error.message}`)
  }
}

export async function updateAnalysisResult(
  id: string,
  result: AnalysisResult
): Promise<void> {
  const { error } = await getSupabaseAdmin()
    .from('reports')
    .update({
      analysis_result: result,
      status: 'complete',
      completed_at: new Date().toISOString(),
    })
    .eq('id', id)

  if (error) {
    console.error('updateAnalysisResult error:', error)
    throw new Error(`Failed to update analysis result: ${error.message}`)
  }
}

export async function incrementShareCount(id: string): Promise<void> {
  const { error } = await getSupabaseAdmin().rpc('increment_share_count', {
    report_id: id,
  })

  if (error) {
    // Fallback: read then write
    const report = await getReport(id)
    if (report) {
      await getSupabaseAdmin()
        .from('reports')
        .update({ share_count: report.shareCount + 1 })
        .eq('id', id)
    }
  }
}

File: backend/app/services/supabase.py
"""Supabase CRUD — port of lib/supabase.ts."""

from __future__ import annotations

from datetime import datetime, timezone
from typing import Any, Optional

from app.dependencies import get_supabase
from app.models.product import AnalysisResult, Report


# ---------------------------------------------------------------------------
# Row ↔ Model conversion
# ---------------------------------------------------------------------------

def _to_report(row: dict[str, Any]) -> Report:
    return Report(
        id=row["id"],
        user_email=row["user_email"],
        product_name=row.get("product_name"),
        product_description=row["product_description"],
        image_url=row.get("image_url"),
        target_price=row.get("target_price"),
        main_concern=row.get("main_concern"),
        quantity=row.get("quantity"),
        analysis_result=row.get("analysis_result"),
        status=row.get("status", "pending"),
        stripe_session_id=row.get("stripe_session_id"),
        stripe_payment_id=row.get("stripe_payment_id"),
        is_paid=row.get("is_paid", False),
        plan=row.get("plan", "single"),
        share_count=row.get("share_count", 0),
        created_at=row.get("created_at"),
        completed_at=row.get("completed_at"),
    )


# ---------------------------------------------------------------------------
# CRUD
# ---------------------------------------------------------------------------

def get_report(report_id: str) -> Optional[Report]:
    sb = get_supabase()
    resp = sb.table("reports").select("*").eq("id", report_id).execute()
    if not resp.data:
        return None
    return _to_report(resp.data[0])


def update_report(report_id: str, data: dict[str, Any]) -> Report:
    sb = get_supabase()
    resp = sb.table("reports").update(data).eq("id", report_id).execute()
    if not resp.data:
        raise Exception(f"Failed to update report {report_id}")
    return _to_report(resp.data[0])


def mark_processing(report_id: str) -> Report:
    return update_report(report_id, {"status": "processing"})


def mark_failed(report_id: str) -> None:
    sb = get_supabase()
    sb.table("reports").update({"status": "failed"}).eq("id", report_id).execute()


def save_analysis(report_id: str, result: dict[str, Any]) -> Report:
    return update_report(report_id, {
        "analysis_result": result,
        "status": "complete",
        "completed_at": datetime.now(timezone.utc).isoformat(),
    })


# ---------------------------------------------------------------------------
# Scraper cache
# ---------------------------------------------------------------------------

def cache_get(source: str, query_hash: str) -> Optional[dict[str, Any]]:
    sb = get_supabase()
    resp = (
        sb.table("scraped_data_cache")
        .select("data")
        .eq("source", source)
        .eq("query_hash", query_hash)
        .gt("expires_at", datetime.now(timezone.utc).isoformat())
        .execute()
    )
    if resp.data:
        return resp.data[0]["data"]
    return None


def cache_set(
    source: str,
    query_hash: str,
    query: str,
    data: dict[str, Any],
    ttl_seconds: int = 86400,
) -> None:
    from datetime import timedelta

    sb = get_supabase()
    expires_at = (datetime.now(timezone.utc) + timedelta(seconds=ttl_seconds)).isoformat()
    sb.table("scraped_data_cache").upsert(
        {
            "source": source,
            "query_hash": query_hash,
            "query": query,
            "data": data,
            "scraped_at": datetime.now(timezone.utc).isoformat(),
            "expires_at": expires_at,
        },
        on_conflict="source,query_hash",
    ).execute()
